File: web/docs-examples/csharp/Examples/AgentToolCalling.cs
// Copyright (c) Microsoft. All rights reserved.

using Prompty.Core;
using Prompty.OpenAI;

namespace DocsExamples.Examples;

/// <summary>
/// Agent with tool calling — define tools, load agent prompty, invoke with tools.
/// </summary>
public static class AgentToolCalling
{
    /// <summary>
    /// Runs an agent loop that can call registered tools until the LLM
    /// returns a final response or max iterations is reached.
    /// </summary>
    public static async Task<object> RunAsync(
        string promptyPath,
        Dictionary<string, Func<string, Task<string>>> tools,
        Dictionary<string, object?>? inputs = null,
        int maxIterations = 10)
    {
        // One-time setup
        new PromptyBuilder()
            .AddOpenAI();

        // Register built-in tool handlers (function, mcp, etc.)
        ToolDispatch.RegisterBuiltins();

        // Load the agent
        var agent = PromptyLoader.Load(promptyPath);

        // Run the agent loop with tool dispatch
        var result = await Pipeline.InvokeAgentAsync(
            agent,
            inputs,
            tools: tools,
            maxIterations: maxIterations);

        return result;
    }

    /// <summary>
    /// Runs an agent using the [Tool] attribute for automatic tool discovery.
    /// </summary>
    public static async Task<object> RunWithToolAttributeAsync(
        string promptyPath,
        object toolInstance,
        Dictionary<string, object?>? inputs = null,
        int maxIterations = 10)
    {
        // One-time setup
        new PromptyBuilder()
            .AddOpenAI();
        ToolDispatch.RegisterBuiltins();

        // Load and bind tools
        var agent = PromptyLoader.Load(promptyPath);
        var tools = ToolAttribute.BindTools(agent, toolInstance);

        var result = await Pipeline.InvokeAgentAsync(
            agent,
            inputs,
            tools: tools,
            maxIterations: maxIterations);

        return result;
    }
}

/// <summary>
/// Example tool class using the [Tool] attribute for auto-discovery.
/// </summary>
public class WeatherTools
{
    [ToolAttribute(Name = "get_weather", Description = "Get the current weather for a city")]
    public string GetWeather(string city)
    {
        return $"72°F and sunny in {city}";
    }
}

File: web/docs-examples/rust/examples/agent_tool_calling.rs
use prompty::{AgentEvent, TurnOptions};
use serde_json::json;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    prompty::register_defaults();
    prompty_openai::register();

    // Register tool handlers
    prompty::register_tool("get_weather", |args: serde_json::Value| {
        Box::pin(async move {
            let city = args["city"].as_str().unwrap_or("unknown");
            Ok(format!("72°F and sunny in {city}"))
        })
    });

    prompty::register_tool("get_time", |args: serde_json::Value| {
        Box::pin(async move {
            let timezone = args["timezone"].as_str().unwrap_or("UTC");
            Ok(format!("2025-01-15T10:30:00 {timezone}"))
        })
    });

    // Load agent and run with tool-calling loop
    let agent = prompty::load("weather_agent.prompty")?;

    let options = TurnOptions::builder()
        .max_iterations(10)
        .on_event(Box::new(|event| {
            if let AgentEvent::Status(message) = event {
                println!("Agent status: {message}");
            }
        }))
        .build();

    let result = prompty::turn(
        &agent,
        Some(&json!({ "question": "What's the weather in Seattle?" })),
        Some(options),
    )
    .await?;

    println!("Result: {result}");
    Ok(())
}
